### PR TITLE
[core][dashboard] Task backend GC policy - worker update [1/3] 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
Right now, we dropped task status events (e.g. task running, finished, errors) and task profile events (used by ray timeline) on the worker side if there are too many such events accumulated. One issue with the current implementation is data loss occurred at individual event level (e.g. even for which task becomes running). A better semantic is to have data loss at the task attempt level. 

See https://docs.google.com/document/d/1BB2VX9nYPyg80fp5-IokSfWr8vBxxiCkpTinB05QmWw/edit#heading=h.vg3rhzht7moa

This is based on https://github.com/ray-project/ray/pull/34896 which was merged but reverted due to minor perf regression. Rerunning the tests again. 

This PR refactors the worker side events reporting in various ways: 
1. It tracks the task attempts which have data loss, and reports dropped task attempts to GCS. This is so that GCS has enough information of task attempts being dropped on the worker so it could report data loss at task attempt level.
2. It separates task status events from profile events, since the two have different usage and implications for completeness:
   - If some profiling events are missing, it's fine to just report the loss number since each profile event is independent. While status events have dependency between each other
   - profiling events are for timeline features, while the task status events are used to populate both the progress bar and the task table, which needs to be more accurate for users to debug. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rickyyx** [comment]
Ping for review

**Comment by rickyyx** [comment]
Ping for review @rkooo567

**Comment by rkooo567** [comment]
Sorry had to focus on https://github.com/ray-project/ray/pull/40083 today. ETA is tomorrow

File: src/ray/common/ray_config_def.h
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:497]: Hmm this name is confusing to me. Can you explain to me what this exactly means?

File: src/ray/common/ray_config_def.h
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:513]:
```suggestion
RAY_CONFIG(int64_t, task_events_dropped_task_attempt_batch_size, 10 * 1000)
```

Also can you explain why this needs to be batched? (because we should report task attempt to GCS for correct accounting)

Comment (rkooo567) [reply src/ray/common/ray_config_def.h:513]: Also this means each report size will increased. Would this be small enough? (Can you calculate it?)

File: src/ray/core_worker/task_event_buffer.h
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.h:396]:
```suggestion
  absl::flat_hash_set<TaskAttempt> dropped_task_attempts_unreported_
```
?

File: src/ray/core_worker/task_event_buffer.h
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.h:306]: is this necessary requirements? I feel like this method is too complicated due to this featuare. Why don't we just send profile events FIFO? (IIUC this is the status quo?)

Comment (rkooo567) [reply src/ray/core_worker/task_event_buffer.h:306]: I assume it is the optimization, but is it necessary? (feel like GCS can just dedup eventually as it knows all dropped task attempts?)

File: src/ray/core_worker/task_event_buffer.cc
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.cc:494]: Same question as below, but is this necessary for performance? Can we just not do this and handle this in GCS?

**Review by rkooo567 — COMMENTED** [review]
Will take a look at reviewing tests tmrw. Please bear with me if I am talking about something we discussed before (I forgot lol)

File: src/ray/common/ray_config_def.h
Comment (rickyyx) [reply src/ray/common/ray_config_def.h:497]: So there are two kinds:
1. profile events
2. status events.

This controls how many of these to be stored before dropping on the worker.

File: src/ray/common/ray_config_def.h
Comment (rickyyx) [reply src/ray/common/ray_config_def.h:513]: > Also can you explain why this needs to be batched? 

We will need to let GCS know what have been dropped on the worker (otherwise GCS will not be able to do accounting properly) 

> Also this means each report size will increased. Would this be small enough? (Can you calculate it?)

This is exactly the reason why it's batched actually, so 10K of (task id + attempt number) => 10K of (24B + 8B)  => 10K * 32B ~=> 320KiB

File: src/ray/core_worker/task_event_buffer.h
Comment (rickyyx) [reply src/ray/core_worker/task_event_buffer.h:306]: > Why don't we just send profile events FIFO? (IIUC this is the status quo?)

Yeah, we could do this.

File: src/ray/core_worker/task_event_buffer.cc
Comment (rickyyx) [reply src/ray/core_worker/task_event_buffer.cc:494]: We could - but this would result in more "wasted" data being sent. 

A single task attempt could generate:
- 8 status change updates + 8 profile events 

If one of the status change event gets dropped on the worker, the entire task should be dropped and GCed since we are reporting data loss at task attempt level now. 

Sending/recording the remaining 15 updates should probably be avoided if it doesn't make things too complicated?

File: src/ray/core_worker/task_event_buffer.cc
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.cc:410]: I think this status is actually always ok. Can you jsut add assert here?

**Review by rkooo567 — APPROVED** [review]
LGTM. For testing, can you make sure you have the following tests?

1. Test profile events per task is respected.
2. Test max events per status / profiling is respected properly.
3. task_events_dropped_task_attempt_batch_size is respected

**Comment by rkooo567** [comment]
and before merging, please make sure there's no regression in microbenchmark!

**Comment by rickyyx** [comment]
> LGTM. For testing, can you make sure you have the following tests?
> 
> 1. Test profile events per task is respected.
> 2. Test max events per status / profiling is respected properly.
> 3. task_events_dropped_task_attempt_batch_size is respected

Yeah, these are covered in the unit testing: 
1. TestLimitProfileEventsPerTask
2. TestBufferSizeLimit
3. TestBatchedSend

Running microbenchmark + stress test many tasks on the PR.

**Comment by rickyyx** [comment]
Release test signal loooks good - waiting for ci to merge:

Metric                                               | Master           | PR             
----------------------------------------------------|------------------|-----------------
single_client_get_calls_Plasma_Store                | 7282.77          | 7841.54        
single_client_put_calls_Plasma_Store                | 5756.89          | 5737.71        
multi_client_put_calls_Plasma_Store                 | 12672.54         | 12315.34       
single_client_put_gigabytes                         | 18.45            | 19.93          
single_client_tasks_and_get_batch                   | 8.77             | 10.02          
multi_client_put_gigabytes                          | 33.06            | 30.71          
single_client_get_object_containing_10k_refs        | 12.29            | 13.72          
single_client_wait_1k_refs                          | 5.30             | 5.56           
single_client_tasks_sync                            | 1218.19          | 1167.51        
single_client_tasks_async                           | 9753.34          | 9949.97        
multi_client_tasks_async                            | 27750.93         | 29401.67       
1_1_actor_calls_sync                                | 2238.65          | 2212.49        
1_1_actor_calls_async                               | 7546.64          | 7497.98        
1_1_actor_calls_concurrent                          | 4798.26          | 4503.25        
1_n_actor_calls_async                               | 9732.13          | 9250.60        
n_n_actor_calls_async                               | 30251.36         | 30422.05       
n_n_actor_calls_with_arg_async                      | 2925.57          | 2925.14        
1_1_async_actor_calls_sync                          | 1451.97          | 1479.63        
1_1_async_actor_calls_async                         | 3164.91          | 3027.01        
1_1_async_actor_calls_with_args_async               | 2329.22          | 2260.64        
1_n_async_actor_calls_async                         | 8936.29          | 8174.14        
n_n_async_actor_calls_async                         | 25270.14         | 25046.85       
placement_group_create/removal                      | 1012.41          | 1002.06        
client__get_calls                                   | 1260.04          | 1141.05        
client__put_calls                                   | 928.00           | 958.77         
client__put_gigabytes                               | 0.13             | 0.09           
client__tasks_and_put_batch                         | 11604.27         | 12290.04       
client__1_1_actor_calls_sync                        | 569.51           | 558.57         
client__1_1_actor_calls_async                       | 1051.85          | 1008.41        
client__1_1_actor_calls_concurrent                  | 1041.03          | 978.42         
client__tasks_and_get_batch                         | 0.98             | 1.06

**Pull request merged**
